src/geocel/rasterize/ImageWriter.libpng.cc

```diff
@@ -0,0 +1,196 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2024 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file geocel/rasterize/ImageWriter.libpng.cc
+//---------------------------------------------------------------------------//
+#include "ImageWriter.hh"
+
+#include <csetjmp>
+#include <cstdio>
+#include <cstring>
+#include <png.h>
+
+#include "corecel/io/Logger.hh"
+
+namespace celeritas
+{
+namespace
+{
+//---------------------------------------------------------------------------//
+struct FileDeleter
+{
+    void operator()(std::FILE* ptr)
+    {
+        auto result = std::fclose(ptr);
+        if (result)
+        {
+            CELER_LOG(error)
+                << "Failed to close PNG file: " << std::strerror(errno);
+        }
+    }
+};
+
+void log_png_warning(png_structp, png_const_charp msg)
+{
+    CELER_LOG(warning) << msg;
+}
+
+void log_png_error(png_structp png_ptr, png_const_charp msg)
+{
+    CELER_LOG(error) << msg;
+    std::longjmp(png_jmpbuf(png_ptr), 1);
+}
+
+//---------------------------------------------------------------------------//
+}  // namespace
+
+//---------------------------------------------------------------------------//
+struct ImageWriter::Impl
+{
+    std::unique_ptr<std::FILE, FileDeleter> file{nullptr};
+    png_structp png{nullptr};
+    png_infop info{nullptr};
+};
+
+//---------------------------------------------------------------------------//
+//! Deleter for PIMPL idiom
+void ImageWriter::ImplDeleter::operator()(Impl* ptr)
+{
+    delete ptr;
+}
```

Comment:
Yeah, the only reason I did it this way was to allow move construction/assignment, but that's not really something we need. I'll add the `CELER_DELETE_COPY_MOVE` macro and move this there.

Reply:
Actually Windows appears not to like default deleter for undefined class: https://github.com/celeritas-project/celeritas/actions/runs/10111931465/job/27964893847
```
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.40.33807\include\memory(3299): error C2027: use of undefined type 'celeritas::ImageWriter::Impl'
D:\a\celeritas\celeritas\src\geocel/rasterize/ImageWriter.hh(58): note: see declaration of 'celeritas::ImageWriter::Impl'
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.40.33807\include\memory(3299): note: the template instantiation context (the oldest one first) is
D:\a\celeritas\celeritas\src\geocel/rasterize/ImageWriter.hh(60): note: see reference to class template instantiation 'std::unique_ptr<celeritas::ImageWriter::Impl,std::default_delete<celeritas::ImageWriter::Impl>>' being compiled
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.40.33807\include\memory(3335): note: see reference to class template instantiation 'std::default_delete<celeritas::ImageWriter::Impl>' being compiled
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.40.33807\include\memory(3298): note: while compiling class template member function 'void std::default_delete<celeritas::ImageWriter::Impl>::operator ()(_Ty *) noexcept const'
        with
        [
            _Ty=celeritas::ImageWriter::Impl
        ]
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.40.33807\include\memory(3410): note: see the first reference to 'std::default_delete<celeritas::ImageWriter::Impl>::operator ()' in 'std::unique_ptr<celeritas::ImageWriter::Impl,std::default_delete<celeritas::ImageWriter::Impl>>::~unique_ptr'
D:\a\celeritas\celeritas\src\geocel/rasterize/ImageWriter.hh(73): note: see the first reference to 'std::unique_ptr<celeritas::ImageWriter::Impl,std::default_delete<celeritas::ImageWriter::Impl>>::~unique_ptr' in 'celeritas::ImageWriter::ImageWriter'
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.40.33807\include\memory(3299): error C2338: static_assert failed: 'can't delete an incomplete type'
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.40.33807\include\memory(3300): warning C4150: deletion of pointer to incomplete type 'celeritas::ImageWriter::Impl'; no destructor called
D:\a\celeritas\celeritas\src\geocel/rasterize/ImageWriter.hh(58): note: see declaration of 'celeritas::ImageWriter::Impl'
```

Reply:
Good point about the move operations. It fails because you need a complete type, i.e. add `struct ImageWriter::Impl{};` with the other empty definitions in the header.  It is probably fine to keep it as is now.

Reply:
Right duh, it was coming from the inline destructor.